src/software/backend/simulation/simulator_ball.cpp

```diff
@@ -1,60 +1,30 @@
 #include "software/backend/simulation/simulator_ball.h"
 
-std::weak_ptr<PhysicsBall> SimulatorBallSingleton::physics_ball_weak_ptr =
-    std::weak_ptr<PhysicsBall>();
+#include "software/logger/init.h"
 
-void SimulatorBallSingleton::setPhysicsBall(std::weak_ptr<PhysicsBall> ball)
+SimulatorBall::SimulatorBall(std::weak_ptr<PhysicsBall> physics_ball)
+    : physics_ball(physics_ball)
 {
-    physics_ball_weak_ptr = ball;
 }
 
-std::unique_ptr<FirmwareBall_t, FirmwareBallDeleter>
-SimulatorBallSingleton::createFirmwareBall()
+Point SimulatorBall::position() const
 {
-    // TODO: Make sure all objects de-allocated properly
-    // See issue https://github.com/UBC-Thunderbots/Software/issues/1128
-    FirmwareBall_t* firmware_ball =
-        app_firmware_ball_create(&(SimulatorBallSingleton::getBallPositionX),
-                                 &(SimulatorBallSingleton::getBallPositionY),
-                                 &(SimulatorBallSingleton::getBallVelocityX),
-                                 &(SimulatorBallSingleton::getBallVelocityY));
-
-    return std::unique_ptr<FirmwareBall_t, FirmwareBallDeleter>(firmware_ball,
-                                                                FirmwareBallDeleter());
-}
-
-float SimulatorBallSingleton::getBallPositionX()
-{
-    if (auto physics_ball = physics_ball_weak_ptr.lock())
+    if (auto ball = physics_ball.lock())
     {
-        // TODO: implement me
+        return ball->getBallWithTimestamp(Timestamp::fromSeconds(0)).position();
     }
-    return 0.0;
+    LOG(WARNING) << "SimulatorBall being used with invalid PhysicsBall" << std::endl;
+    return Point(0, 0);
 }
 
-float SimulatorBallSingleton::getBallPositionY()
-{
-    if (auto physics_ball = physics_ball_weak_ptr.lock())
-    {
-        // TODO: implement me
-    }
-    return 0.0;
-}
 
-float SimulatorBallSingleton::getBallVelocityX()
-{
-    if (auto physics_ball = physics_ball_weak_ptr.lock())
-    {
-        // TODO: implement me
-    }
-    return 0.0;
-}
 
-float SimulatorBallSingleton::getBallVelocityY()
+Vector SimulatorBall::velocity() const
 {
-    if (auto physics_ball = physics_ball_weak_ptr.lock())
+    if (auto ball = physics_ball.lock())
     {
-        // TODO: implement me
+        return ball->getBallWithTimestamp(Timestamp::fromSeconds(0)).velocity();
     }
-    return 0.0;
+    LOG(WARNING) << "SimulatorBall being used with invalid PhysicsBall" << std::endl;
```

Comment:
Similar comment here, are we sure if we get here it's invalid?

Reply:
Yes. If the weak_ptr is expired (memory may or may not be fully de-allocated yet), or the weak_ptr is empty, `lock()` returns an empty pointer. In either case we don't have a valid object to operate on.

https://en.cppreference.com/w/cpp/memory/weak_ptr/lock